docs/software_requirements/semaphore.sdoc

```diff
@@ -0,0 +1,245 @@
+[DOCUMENT]
+TITLE: Semaphores
+REQ_PREFIX: ZEP-
+
+[GRAMMAR]
+IMPORT_FROM_FILE: software_requirements.sgra
+
+[REQUIREMENT]
+UID: ZEP-12001
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Counting Semaphore Definition At Compile Time
+STATEMENT: >>>
+The Zephyr RTOS shall provide a mechanism to define and initialize a semaphore at compile time.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-99
+
+[REQUIREMENT]
+UID: ZEP-12002
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Counting Semaphore Definition At Run Time
+STATEMENT: >>>
+The Zephyr RTOS shall provide a mechanism to define and initialize a semaphore at runtime.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-99
+
+[REQUIREMENT]
+UID: ZEP-12003
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Maximum limit of a semaphore
+STATEMENT: >>>
+The Zephyr RTOS shall define the maximum limit of a semaphore when the semaphore is used for counting purposes and does not have an explicit limit.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-12004
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Initialialization with maximum count value
+STATEMENT: >>>
+When initializing a counting semaphore, the maximum permitted count a semaphore
+can have shall be set.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-99
+
+[REQUIREMENT]
+UID: ZEP-12005
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Initialialization with initial semaphore value
+STATEMENT: >>>
+When initializing a counting semaphore, the initial semaphore value shall be set.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-99
+
+[REQUIREMENT]
+UID: ZEP-12006
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Semaphore acquisition mechanism
+STATEMENT: >>>
+The Zephyr RTOS shall provide a mechanism allowing threads to acquire a semaphore.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-99
+
+[REQUIREMENT]
+UID: ZEP-12007
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Semaphore acquisition with count greater than zero
+STATEMENT: >>>
+If the semaphore's count is greater than zero, the requesting thread shall acquire
+the semaphore and decrement its count.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-99
+
+[REQUIREMENT]
+UID: ZEP-12008
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Semaphore acquisition with zero count
+STATEMENT: >>>
+If the semaphore's count is zero, the requesting thread shall be blocked until the semaphore is released by another thread.
+<<<
```

Comment:
1. "Until" prevents this statement from being Atomic.
2. "If" as used here will create confusion with how it is used in the recommended EARS syntax, where it indicates an undesirable condition.
3. This behavior does not apply under all conditions, so the statement is missing at least one conditional. One case where this behavior does not apply is when it is requested as a non-blocking call. Until these conditions are stated here the requirement is not Independent of other requirements.

Consider expressing it in the (recommended) EARS syntax. Here is an example:
While a Counting Semaphore count is 0,

When a thread takes that Counting Semaphore with a FOREVER timeout,

the Zephyr RTOS shall block the taking thread waiting for the availability of that Counting Semaphore.